### PR TITLE
Lynn-2: Intersector

### PR DESCRIPTION
**Pull request by HaiDang2001VN**
Upload all needed files (headers and source).
Usage: re-define macro according to yours OS (originally Windows).

### CODE REVIEW

**Comment by lamnhh** [comment]
Could you describe your PR in more details (what it is and what it does)?

**Comment by HaiDang2001VN** [comment]
@lamnhh thanks for asking. This PR is supposed to project a point (type `cv::Vec3f`) or a polygon (type `vector<cv::Vec3f>`) to a given plane with given center. For explicit explanation, given a pyramid whose bottom is polygon, calculate the cross section (a polygon) of it with given plane.

File: src/Intersection.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
+﻿// Intersection.cpp : This file contains the 'main' function. Program execution begins and ends there.
+//
+
+#include "pch.h"
+#include "Intersector.h"
+#include <iostream>
+
+using namespace std;
+
+float a, b, c, d, x, y, z, m, n, o = 0;
```
Comment (lamnhh) [inline src/Intersection.cpp]: Please avoid using global variables.

File: src/Intersector.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#include "pch.h"
+
+#ifdef LINUX
+#include "Intersector.h"
+#endif
+
+#define sqr(x) x * x
```
Comment (lamnhh) [inline src/Intersector.cpp]: Remember to #undef it later.

**Comment by lamnhh** [comment]
@HaiDang2001VN Thanks for the answer. I've checked the formula in the code and it's indeed correct. However, I propose a simpler formula like this:
Consider we are computing the intersection between ray OP and plane TarPlane: ax + by + cz + d = 0. Assume coordinates for O and P are (x0, y0, z0) and (x1, y1, z1) respectively.
Define vx = x1 - x0, vy = y1 - y0, vz = z1 - z0.
We have a system:
```
x = x0 + vx * t
y = y0 + vy * t
z = z0 + vz * t
ax + by + cz + d = 0
```
Substitute x, y, z with regard to t into the fourth equation, we have:
```
t = -(ax0 + by0 + cz0 + d) / (avx + bvy + cvz);
ans = O + t * v;
```
I believe this is more intuitive and less computational expensive.
Also, what are the empty Estimate() functions in the end of "Intersector.cpp"?

File: src/Intersection.cpp (hunk 1 above)
Comment (HaiDang2001VN) [reply src/Intersection.cpp]: @lamnhh Thanks for pointing out that, this variables are used in testing only.

File: src/Intersector.cpp (hunk 2 above)
Comment (HaiDang2001VN) [reply src/Intersector.cpp]: @lamnhh Could you please explain why we have to #undef it, I hasn't known about it until now.

**Comment by HaiDang2001VN** [comment]
@lamnhh thanks for proposing new formula!
After some comparisons, I figured out that except for `-(ax0 + by0 + cz0 + d) / (avx + bvy + cvz)` part, the rests is the same with my code as I wrote it in `cv::Vec` form with some dot products.
About the difference, I find yours more inuitive and simpler to compute in this case because of algerabic form of inputs.
So, I will change that. Thanks for your proposal.

**Comment by HaiDang2001VN** [comment]
@lamnhh 
About the empty functions, in the begining I planed to explore how to compute the matrix form of this task and provide transformation matrix as original purpose (which will result in transformed point after dot with this matrix). I left these because I have been unable to find any way making it possible.
So I hope someone in the future can solve this problem.

File: src/Intersection.cpp (hunk 1 above)
Comment (phamdat1992) [reply src/Intersection.cpp]: you should move the testing code to test/integration folder.

**Comment by HaiDang2001VN** [comment]
Thanks @lamnhh  and @phamdat1992  for reviewing my code.
I have changed code according to suggestion of @lamnhh  and refactor the test on @phamdat1992 's demand. Please have a check on my code. Thanks!

**Comment by lamnhh** [comment]
@HaiDang2001VN Please undef the sqr(x) directive. It may cause undesired increase in complexity in the future. You can look at this code to see why: https://ideone.com/kUYupI
Or you can write it as a function. Declare it as static to encapsulate it inside this file.

**Comment by HaiDang2001VN** [comment]
@lamnhh  thanks for your review, I have changed it to remove the risk. Please check it out.

**Review by lamnhh — APPROVED** [review]

**Pull request merged**